streaming/src/event_server.h

```diff
@@ -0,0 +1,216 @@
+
+/// The event queue inherits from the universal ring queue. It mainly divides
+/// event into two different levels: normal event and urgent event, and the
+/// total size of the queue is the sum of them.
+class EventQueue : public AbstractRingBuffer<Event> {
+ public:
+  EventQueue(size_t size) : urgent_(false), capacity_(size), is_started_(true) {}
+
+  virtual ~EventQueue() {
+    is_started_ = false;
+    full_cv_.notify_all();
+    empty_cv_.notify_all();
+  };
+
+  void Run() { is_started_ = true; }
+
+  void Stop() {
+    is_started_ = false;
+    empty_cv_.notify_all();
+    full_cv_.notify_all();
+  }
+
+  void Push(const Event &t) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Size() >= capacity_ && is_started_) {
+      STREAMING_LOG(WARNING) << " EventQueue is full, its size:" << Size()
+                             << " all_size:" << Size() << " capacity:" << capacity_
+                             << " buffer size:" << buffer_.size()
+                             << " urgent_buffer size:" << urgent_buffer_.size();
+      full_cv_.wait(lock);
+      STREAMING_LOG(WARNING) << "Event server is full_sleep be notified";
+    }
+    if (!is_started_) {
+      return;
+    }
+    if (t.urgent) {
+      buffer_.push(t);
+    } else {
+      urgent_buffer_.push(t);
+    }
+    if (1 == Size()) {
+      empty_cv_.notify_one();
+    }
+  }
+
+  void Pop() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    if (Size() >= capacity_) {
+      STREAMING_LOG(WARNING) << "Pop should notify"
+                             << " all_Size:" << Size();
+    }
+    if (urgent_) {
+      urgent_buffer_.pop();
+    } else {
+      buffer_.pop();
+    }
+    full_cv_.notify_all();
+  }
+
+  bool Get(Event &evt) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return false;
+    }
+    if (!urgent_buffer_.empty()) {
+      urgent_ = true;
+      evt = urgent_buffer_.front();
+    } else {
+      urgent_ = false;
+      evt = buffer_.front();
+    }
+    return true;
+  }
+
+  Event PopAndGet() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return Event();
+    }
+    if (!urgent_buffer_.empty()) {
+      Event res = urgent_buffer_.front();
+      urgent_buffer_.pop();
+      if (Full()) {
+        full_cv_.notify_one();
+      }
+      return res;
+    }
+    Event res = buffer_.front();
+    buffer_.pop();
+    if (Size() + 1 == capacity_) full_cv_.notify_one();
+    return res;
+  }
+
+  Event &Front() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    if (urgent_buffer_.size()) {
+      return urgent_buffer_.front();
+    }
+    return buffer_.front();
+  }
+
+  inline bool Empty() const { return buffer_.empty() && urgent_buffer_.empty(); }
```

Comment:
These tow methods are not atomic. After `urgent_buffer_.empty()` return empty, `buffer_ ` may not be empty.

Reply:
> These tow methods are not atomic. After `urgent_buffer_.empty()` return empty, `buffer_ ` may not be empty.

It's enough to keep thread-safe in push/pop function and no-strict safe is acceptable to other usages.